### ISSUE TITLE
atom type values higher than 1 when resolution is 1

### ISSUE THREAD

**Issue by mgm-14392** [gnina/libmolgrid#67]
Hello, I understand that "atom type information is represented as a density distribution around the atom center" and I believe the function you are using is the one presented in this paper: https://doi.org/10.1021/acs.jcim.6b00740. Therefore, I'd expect the atom type values to be constrained between 0 and 1. This is true when the GridMaker has a resolution=0.5 and dimension=23.5. When I change to resolution=1.0 and dimension=23, I observe that some of the values in the grid/tensor are above 1.

When the resolution is 0.5, the max value is 1. When I change the resolution to 1.0 the max values are higher than one. Am I doing something wrong?

Thank you!

Here is a small example, my types file had 10 examples:

    import molgrid
    import torch

    datadir = "data"
    dname = "check_voxel.types"

    d = molgrid.ExampleProvider(shuffle=False, balanced=False,
                                stratify_receptor=False, data_root=datadir)
    d.populate(dname)
    
    gmaker = molgrid.GridMaker(resolution=1.0, dimension=23, binary=False,
                               radius_type_indexed=False, radius_scale=1.0, gaussian_radius_multiple=1.0)
    gmaker2 = molgrid.GridMaker(resolution=0.5, dimension=23.5, binary=True,
                               radius_type_indexed=False, radius_scale=1.0, gaussian_radius_multiple=1.0)

    ddims = gmaker.grid_dimensions(d.num_types())
    ddims2 = gmaker2.grid_dimensions(d.num_types())

    batch_size = 1
    dtensor_shape = (batch_size,) + ddims
    dtensor_shape2 = (batch_size,) + ddims2

    dinput_tensor = torch.zeros(dtensor_shape, dtype=torch.float32)
    dinput_tensor2 = torch.zeros(dtensor_shape2, dtype=torch.float32)

    for i in range(0,10):
        dbatch = d.next_batch(batch_size)
        gmaker.forward(dbatch, dinput_tensor, random_translation=0.0, random_rotation=False)
        gmaker2.forward(dbatch, dinput_tensor2, random_translation=0.0, random_rotation=False)

        print('res 1.0')
        print(torch.max(dinput_tensor))
        print('res 0.5')
        print(torch.max(dinput_tensor2))

Example outputs are:

    res 1.0
    tensor(2.0447)
    res 0.5
    tensor(1.)
    res 1.0
    tensor(2.0685)
    res 0.5
    tensor(1.)

**Comment by dkoes** [comment]
If atom densities are overlapping you will get density values greater than one.  

Your code has binary set to True for gmaker2 but not gmaker, which is probably why you are seeing these differences (binary output does not accumulate).   Please reopen if changing the binary setting doesn't produce sensible outputs.

**Comment by mgm-14392** [comment]
Thank you! True, I made a mistake there. Is there anyway I could normalize the values other than using binary?

**Comment by dkoes** [comment]
Use torch operations to achieve your desired normalization.
